### PR TITLE
Several FARMS radiation related bug fixes

### PR DESCRIPTION
**Pull request by davegill**
TYPE: bug fix

KEYWORDS: FARMS, radii, Goddard

SOURCE: internal

DESCRIPTION OF CHANGES:
Problem:
The FARMS radiation option (`swint_opt==2`) assumes both the availabilty of effective radii
and mass for three species: cloud, ice, and snow. This causes several issues.
1. The Goddard MP scheme had the effective radii in a separate variable, in different units.
2. The effective radii for snow could be valid with an unallocated mass for snow (for example, WSM3).
3. Two locations in radiation driver that need to have the FARMS option in the IF test
4. Even though use_mp_re=1 is a default, if someone turns it off, the FARMS option is not possible 
(but that was not trapped).

Solution:
1. In the radiation driver, if the MP scheme is Goddard, translate the radii to the correct
variable name and with the correct units.
2. If the FARMS radiation option (`swint_opt==2`) is used, in the physics initialization
routine require that effective radii and mass for cloud, ice, and snow be available.
3. Welp, just add those additional `.or. swint_opt ==2` in (per Jimy and Pedro).
4. Add an IF test in check_a_mundo that FARMS + use_mp_re=0 is a FATAL error.

LIST OF MODIFIED FILES:
modified:   phys/module_physics_init.F
modified:   phys/module_radiation_driver.F
modified:   share/module_check_a_mundo.F

TESTS CONDUCTED:
1. jenkins is all PASS
2. With acceptable nml (FARMS + MP with sufficient radii and mass info)
```
 &physics
 physics_suite = 'CONUS'
 swint_opt     = 2
 mp_physics    = 7
```
The _temporary_ messages indicate positive results
```
 In physics init, we have the correct radii and masses
INPUT LandUse = "USGS"
 LANDUSE TYPE = "USGS" FOUND          33  CATEGORIES           2  SEASONS WATER CATEGORY =           16  SNOW CATEGORY =           24
INITIALIZE THREE Noah LSM RELATED TABLES
 LANDUSE TYPE = USGS FOUND          27  CATEGORIES
 INPUT SOIL TEXTURE CLASSIFICATION = STAS
 SOIL TEXTURE CLASSIFICATION = STAS FOUND          19  CATEGORIES
Timing for Writing wrfout_d01_2000-01-24_12:00:00 for domain        1:    0.40447 elapsed seconds
d01 2000-01-24_12:00:00  Input data is acceptable to use: wrfbdy_d01
Timing for processing lateral boundary for domain        1:    0.03351 elapsed seconds
 Tile Strategy is not specified. Assuming 1D-Y
WRF TILE   1 IS      1 IE     74 JS      1 JE     61
WRF NUMBER OF TILES =   1
 In radiation driver, converting Goddard MP radii to traditional
```
3. With insufficient mass fields in WSM3 (no snow) to run the FARMS scheme:
```
 &physics
 physics_suite = 'CONUS'
 swint_opt     = 2
 mp_physics    = 3
```
The error is trapped
```
-------------- FATAL CALLED ---------------
FATAL CALLED FROM FILE:  <stdin>  LINE:     911
--- ERROR: FARMS (swint_opt==2) requires a different MP scheme
```
4. If a user manages to turn off `use_mp_re==0` and asks for FARMS (an inconsistent combination)
```
 &physics
 physics_suite = 'CONUS'
 swint_opt     = 2
 use_mp_re     = 0
```
This error combination is successfully trapped
```
--- ERROR: FARMS (swint_opt=2) requires effective radii (use_mp_re=1)
-------------- FATAL CALLED ---------------
FATAL CALLED FROM FILE:  <stdin>  LINE:    2339
NOTE:       1 namelist settings are wrong. Please check and reset these options
```

### CODE REVIEW

File: phys/module_physics_init.F [hunk 1]
```diff
@@ -938,7 +938,22 @@ SUBROUTINE phy_init ( id, config_flags, DT, restart, zfull, zhalf,     &
...
+   IF ( config_flags%swint_opt .eq. swintopt2 ) THEN
+      IF ( ( has_reqc .EQ. 1 ) .AND. &
+           ( has_reqi .EQ. 1 ) .AND. &
+           ( has_reqs .EQ. 1 ) .AND. &
+           ( f_qc            ) .AND. &
+           ( f_qi            ) .AND. &
+           ( f_qs            ) ) THEN
+         !  everything is A-OK for FARMS
+      ELSE
+         CALL wrf_error_fatal ('--- ERROR: FARMS requires a different MP scheme')
```
Comment (weiwangncar) [inline phys/module_physics_init.F:952]: @davegill Could we add (swint_opt=2) after FARMS?

Comment (davegill) [reply phys/module_physics_init.F:953]: Wei,
Good suggestion. I put that in.

**Comment by davegill** [comment]
jenkins
```
Please find result of the WRF regression test cases in the attachment. This build is for Commit ID: 6091c9e12f544b0a4e78759adefe4faf601318d2, requested by: davegill for PR: https://github.com/wrf-model/WRF/pull/1174. For any query please send e-mail to David Gill.

    Test Type              | Expected  | Received |  Failed
    = = = = = = = = = = = = = = = = = = = = = = = =  = = = =
    Number of Tests        : 18           18
    Number of Builds       : 46           46
    Number of Simulations  : 164           164        0
    Number of Comparisons  : 104           104        0

    Failed Simulations are: 
    None
    Which comparisons are not bit-for-bit: 
    None
```

**Comment by davegill** [comment]
@weiwangncar @dudhia @pedro-jm 
Folks,
Please take a look at this. It is mostly a clean-up for FARMS.

**Comment by dudhia** [comment]
A better clue in the error message is it requires a microphysics scheme that provides effective radii. Otherwise looks fine.

**Comment by weiwangncar** [comment]
So solar diagnostics and FARMS are independent options?

**Comment by dudhia** [comment]
solar_diagnostics gives optional extra output with or without FARMS.

**Pull request merged**